StakingRewards/contracts/StakingRewards.sol

```diff
@@ -0,0 +1,306 @@
+  /* -------------------------------------------------------------------------- */
+
+  /**
+   * Emitted when user stakes staking token
+   */
+  event Staked(address indexed account, uint256 amount);
+  /**
+   * Emited when user unstakes his/her staked token
+   */
+  event Unstaked(address indexed account, uint256 amount);
+  /**
+   * Emitted when user claim his/her rewards token
+   */
+  event Claimed(address indexed account, uint256 amount);
+  /**
+   * Emitted when owner funds rewards token and restart rewarding session
+   */
+  event Funded(uint256 amount);
+  /**
+   * Emitted when rewards duration has been updated
+   */
+  event RewardsDurationUpdated(uint256 duration);
+  /**
+   * Emitted when owner recovered tokens from this contract
+   */
+  event Recovered(address token, uint256 amount);
+
+  /* -------------------------------------------------------------------------- */
+  /*                                  Modifier                                  */
+  /* -------------------------------------------------------------------------- */
+
+  modifier updateReward(address account) {
+    rewardPerTokenStored = _rewardPerToken();
+    lastUpdateTime = _lastTimeRewardApplicable();
+    if (account != address(0)) {
+      rewards[account] = _earned(account);
+      userRewardPerTokenPaid[account] = rewardPerTokenStored;
+    }
+    _;
+  }
+
+  /* -------------------------------------------------------------------------- */
+  /*                             External Functions                             */
+  /* -------------------------------------------------------------------------- */
+
+  /**
+   * @dev Constructor
+   * @param rewardsToken_ Address to rewards ERC20 token
+   * @param stakingToken_ Address to staking ERC20 token
+   * @param duration Reward duration in seconds
+   */
+  constructor(address rewardsToken_, address stakingToken_, uint256 duration) {
+    rewardsToken = IERC20(rewardsToken_);
+    stakingToken = IERC20(stakingToken_);
+    rewardsDuration = duration;
+  }
+
+  /**
+   * @notice Set rewards duration, only available to set after finish
+   * of previous rewards period.
+   * @dev Callable by owner
+   * @param duration New rewards duration
+   */
+  function setRewardsDuration(uint256 duration) external onlyOwner {
+    if (block.timestamp < periodFinish) {
+      revert WaitToFinish();
+    }
+    rewardsDuration = duration;
+    emit RewardsDurationUpdated(duration);
+  }
+
+  /**
+   * @notice Stake staking token, callable only when unpaused.
+   * Every staked balance per user will be accumulated and it transfers tokens
+   * from user to this contract.
+   * @dev Callable only when unpaused
+   * @param amount Staking amount
+   */
+  function stake(
+    uint256 amount
+  ) public nonReentrant whenNotPaused updateReward(msg.sender) {
+    if (amount <= 0) {
+      revert ZeroAmount();
+    }
+    totalSupply = totalSupply + amount;
```

Comment:
`totalSupply += amount;`

Reply:
`totalSupply = totalSupply + amount;` is more gasless